test/hummingbot/connector/exchange/tokocrypto/test_tokocrypto_api_order_book_data_source.py

```diff
@@ -0,0 +1,413 @@
+        url = web_utils.public_rest_url(path_url=CONSTANTS.SNAPSHOT_PATH_URL, domain=self.domain)
+        regex_url = re.compile(f"^{url}".replace(".", r"\.").replace("?", r"\?"))
+
+        mock_api.get(regex_url, status=400)
+        with self.assertRaises(IOError):
+            self.async_run_with_timeout(
+                self.data_source.get_new_order_book(self.trading_pair)
+            )
+
+    @patch("aiohttp.ClientSession.ws_connect", new_callable=AsyncMock)
+    def test_listen_for_subscriptions_subscribes_to_trades_and_order_diffs(self, ws_connect_mock):
+        ws_connect_mock.return_value = self.mocking_assistant.create_websocket_mock()
+
+        result_subscribe_trades = {
+            "result": None,
+            "id": 1
+        }
+        result_subscribe_diffs = {
+            "result": None,
+            "id": 2
+        }
+
+        self.mocking_assistant.add_websocket_aiohttp_message(
+            websocket_mock=ws_connect_mock.return_value,
+            message=json.dumps(result_subscribe_trades))
+        self.mocking_assistant.add_websocket_aiohttp_message(
+            websocket_mock=ws_connect_mock.return_value,
+            message=json.dumps(result_subscribe_diffs))
+
+        self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_subscriptions())
+
+        self.mocking_assistant.run_until_all_aiohttp_messages_delivered(ws_connect_mock.return_value)
+
+        sent_subscription_messages = self.mocking_assistant.json_messages_sent_through_websocket(
+            websocket_mock=ws_connect_mock.return_value)
+
+        self.assertEqual(2, len(sent_subscription_messages))
+        expected_trade_subscription = {
+            "method": "SUBSCRIBE",
+            "params": [f"{self.ex_trading_pair.lower()}@trade"],
+            "id": 1}
+        self.assertEqual(expected_trade_subscription, sent_subscription_messages[0])
+        expected_diff_subscription = {
+            "method": "SUBSCRIBE",
+            "params": [f"{self.ex_trading_pair.lower()}@depth@100ms"],
+            "id": 2}
+        self.assertEqual(expected_diff_subscription, sent_subscription_messages[1])
+
+        self.assertTrue(self._is_logged(
+            "INFO",
+            "Subscribed to public order book and trade channels..."
+        ))
+
+    @patch("hummingbot.core.data_type.order_book_tracker_data_source.OrderBookTrackerDataSource._sleep")
+    @patch("aiohttp.ClientSession.ws_connect")
+    def test_listen_for_subscriptions_raises_cancel_exception(self, mock_ws, _: AsyncMock):
+        mock_ws.side_effect = asyncio.CancelledError
+
+        with self.assertRaises(asyncio.CancelledError):
+            self.listening_task = self.ev_loop.create_task(self.data_source.listen_for_subscriptions())
+            self.async_run_with_timeout(self.listening_task)
+
+    @patch("hummingbot.core.data_type.order_book_tracker_data_source.OrderBookTrackerDataSource._sleep")
```

Comment:
The patch should be done on top of the Tokocrypto order book data source class, not on top of the super class